### ISSUE TITLE
Templating init failed Minified Redux error #3 when loading dashboards of schema version 31

### ISSUE THREAD

**Issue by Yogu** [grafana/grafana#42943]
**What happened**:

We updated from Grafana 8.2.1 to 8.3.1, and now all dashboards have the following behavior:

- They display the error:
```
Templating init failed Minified Redux error #3; visit https://redux.js.org/Errors?code=3 for the full message or use the non-minified dev environment for full errors.
```

when you load them

```
Error: Minified Redux error #3; visit https://redux.js.org/Errors?code=3 for the full message or use the non-minified dev environment for full errors. 
    at Object.m [as getState] (redux.js:188)
    at r (store.ts:19)
    at Object.o [as getVariableWithName] (selectors.ts:27)
    at Object.getVariableAtIndex (template_srv.ts:338)
    at template_srv.ts:269
    at String.replace (<anonymous>)
    at Object.replace (template_srv.ts:267)
    at d.getInstanceSettings (datasource_srv.ts:87)
    at Object.beforeAdding (adapter.ts:50)
    at addVariable (sharedReducer.ts:19)
```

- All variable dropdowns are empty
- If you open the variables in the settings, they don't show any values

There is two way to fix it:
- Click "Update" on a variable in the settings, which refreshes the values
- Just click "Save", then hit F5.

**What you expected to happen**:

Dashboards should continue to work after the update

**How to reproduce it (as minimally and precisely as possible)**:

Here's one of our dashboards pre-conversion:

```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 150,
  "iteration": 1639066164477,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "datasource": {
        "uid": "${datasource}"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "log": 2,
              "type": "log"
            },
            "showPoints": "never",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          },
          "unit": "bytes"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 24,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "targets": [
        {
          "exemplar": true,
          "expr": "next_cluster_fs_free",
          "hide": false,
          "interval": "",
          "legendFormat": "{{dbhost}}",
          "refId": "A"
        }
      ],
      "title": "Free space on hosts",
      "type": "timeseries"
    },
    {
      "datasource": {
        "uid": "${datasource}"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "align": "auto",
            "displayMode": "auto"
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          },
          "unit": "bytes"
        },
        "overrides": [
          {
            "matcher": {
              "id": "byName",
              "options": "dbhost"
            },
            "properties": [
              {
                "id": "custom.width",
                "value": 395
              }
            ]
          },
          {
            "matcher": {
              "id": "byName",
              "options": "used %"
            },
            "properties": [
              {
                "id": "unit",
                "value": "percentunit"
              }
            ]
          }
        ]
      },
      "gridPos": {
        "h": 8,
        "w": 23,
        "x": 0,
        "y": 9
      },
      "id": 4,
      "options": {
        "footer": {
          "fields": "",
          "reducer": [
            "sum"
          ],
          "show": false
        },
        "frameIndex": 0,
        "showHeader": true,
        "sortBy": [
          {
            "desc": true,
            "displayName": "total"
          }
        ]
      },
      "pluginVersion": "8.3.1",
      "targets": [
        {
          "exemplar": true,
          "expr": "next_cluster_fs_free",
          "format": "table",
          "instant": true,
          "interval": "",
          "legendFormat": "",
          "refId": "A"
        },
        {
          "exemplar": true,
          "expr": "next_cluster_fs_used",
          "format": "table",
          "hide": false,
          "instant": true,
          "interval": "",
          "legendFormat": "",
          "refId": "B"
        },
        {
          "exemplar": true,
          "expr": "next_cluster_fs_total",
          "format": "table",
          "hide": false,
          "instant": true,
          "interval": "",
          "legendFormat": "",
          "refId": "C"
        }
      ],
      "title": "Hosts",
      "transformations": [
        {
          "id": "seriesToColumns",
          "options": {
            "byField": "dbhost"
          }
        },
        {
          "id": "organize",
          "options": {
            "excludeByName": {
              "Time 1": true,
              "Time 2": true,
              "Time 3": true,
              "__name__ 1": true,
              "__name__ 2": true,
              "__name__ 3": true,
              "cluster 1": true,
              "cluster 2": true,
              "cluster 3": true,
              "hostname 1": true,
              "hostname 2": true,
              "hostname 3": true,
              "instance 1": true,
              "instance 2": true,
              "instance 3": true,
              "job 1": true,
              "job 2": true,
              "job 3": true,
              "kubernetes_namespace 1": true,
              "kubernetes_namespace 2": true,
              "kubernetes_namespace 3": true,
              "path 1": true,
              "path 2": true,
              "path 3": true,
              "service_name 1": true,
              "service_name 2": true,
              "service_name 3": true
            },
            "indexByName": {},
            "renameByName": {
              "Value #A": "free",
              "Value #B": "used",
              "Value #C": "total"
            }
          }
        },
        {
          "id": "calculateField",
          "options": {
            "alias": "used %",
            "binary": {
              "left": "used",
              "operator": "/",
              "reducer": "sum",
              "right": "total"
            },
            "mode": "binary",
            "reduce": {
              "include": [
                "free",
                "total"
              ],
              "reducer": "diffperc"
            }
          }
        }
      ],
      "type": "table"
    }
  ],
  "refresh": "",
  "schemaVersion": 33,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": [
      {
        "current": {
          "selected": true,
          "text": "nextprod02apps",
          "value": "nextprod02apps"
        },
        "hide": 0,
        "includeAll": false,
        "multi": false,
        "name": "datasource",
        "options": [],
        "query": "prometheus",
        "queryValue": "",
        "refresh": 1,
        "regex": "next.*apps",
        "skipUrlSync": false,
        "type": "datasource"
      }
    ]
  },
  "time": {
    "from": "now-15m",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Database hosts",
  "uid": "uXEvfqd7k",
  "version": 4,
  "weekStart": ""
}
```

Here is it post-update

```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 150,
  "iteration": 1639066404149,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "datasource": {
        "uid": "${datasource}"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "log": 2,
              "type": "log"
            },
            "showPoints": "never",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          },
          "unit": "bytes"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 24,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "targets": [
        {
          "exemplar": true,
          "expr": "next_cluster_fs_free",
          "hide": false,
          "interval": "",
          "legendFormat": "{{dbhost}}",
          "refId": "A"
        }
      ],
      "title": "Free space on hosts",
      "type": "timeseries"
    },
    {
      "datasource": {
        "uid": "${datasource}"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "align": "auto",
            "displayMode": "auto"
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          },
          "unit": "bytes"
        },
        "overrides": [
          {
            "matcher": {
              "id": "byName",
              "options": "dbhost"
            },
            "properties": [
              {
                "id": "custom.width",
                "value": 395
              }
            ]
          },
          {
            "matcher": {
              "id": "byName",
              "options": "used %"
            },
            "properties": [
              {
                "id": "unit",
                "value": "percentunit"
              }
            ]
          }
        ]
      },
      "gridPos": {
        "h": 8,
        "w": 23,
        "x": 0,
        "y": 9
      },
      "id": 4,
      "options": {
        "footer": {
          "fields": "",
          "reducer": [
            "sum"
          ],
          "show": false
        },
        "frameIndex": 0,
        "showHeader": true,
        "sortBy": [
          {
            "desc": true,
            "displayName": "total"
          }
        ]
      },
      "pluginVersion": "8.3.1",
      "targets": [
        {
          "exemplar": true,
          "expr": "next_cluster_fs_free",
          "format": "table",
          "instant": true,
          "interval": "",
          "legendFormat": "",
          "refId": "A"
        },
        {
          "exemplar": true,
          "expr": "next_cluster_fs_used",
          "format": "table",
          "hide": false,
          "instant": true,
          "interval": "",
          "legendFormat": "",
          "refId": "B"
        },
        {
          "exemplar": true,
          "expr": "next_cluster_fs_total",
          "format": "table",
          "hide": false,
          "instant": true,
          "interval": "",
          "legendFormat": "",
          "refId": "C"
        }
      ],
      "title": "Hosts",
      "transformations": [
        {
          "id": "seriesToColumns",
          "options": {
            "byField": "dbhost"
          }
        },
        {
          "id": "organize",
          "options": {
            "excludeByName": {
              "Time 1": true,
              "Time 2": true,
              "Time 3": true,
              "__name__ 1": true,
              "__name__ 2": true,
              "__name__ 3": true,
              "cluster 1": true,
              "cluster 2": true,
              "cluster 3": true,
              "hostname 1": true,
              "hostname 2": true,
              "hostname 3": true,
              "instance 1": true,
              "instance 2": true,
              "instance 3": true,
              "job 1": true,
              "job 2": true,
              "job 3": true,
              "kubernetes_namespace 1": true,
              "kubernetes_namespace 2": true,
              "kubernetes_namespace 3": true,
              "path 1": true,
              "path 2": true,
              "path 3": true,
              "service_name 1": true,
              "service_name 2": true,
              "service_name 3": true
            },
            "indexByName": {},
            "renameByName": {
              "Value #A": "free",
              "Value #B": "used",
              "Value #C": "total"
            }
          }
        },
        {
          "id": "calculateField",
          "options": {
            "alias": "used %",
            "binary": {
              "left": "used",
              "operator": "/",
              "reducer": "sum",
              "right": "total"
            },
            "mode": "binary",
            "reduce": {
              "include": [
                "free",
                "total"
              ],
              "reducer": "diffperc"
            }
          }
        }
      ],
      "type": "table"
    }
  ],
  "refresh": "",
  "schemaVersion": 33,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": [
      {
        "current": {
          "selected": false,
          "text": "nextprod02apps",
          "value": "nextprod02apps"
        },
        "hide": 0,
        "includeAll": false,
        "multi": false,
        "name": "datasource",
        "options": [],
        "query": "prometheus",
        "queryValue": "",
        "refresh": 1,
        "regex": "next.*apps",
        "skipUrlSync": false,
        "type": "datasource"
      }
    ]
  },
  "time": {
    "from": "now-15m",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Database hosts",
  "uid": "uXEvfqd7k",
  "version": 7,
  "weekStart": ""
}
```

Here is the visual diff:
![image](https://user-images.githubusercontent.com/395810/145433818-eeede70f-58a9-4146-aa97-9801c9b51e25.png)

**Anything else we need to know?**:

Seems only to happen for elasticsearch data sources

WebSocket connections fail, probably due to our reverse proxy, but I don't think this is related.

**Environment**:
- Grafana version: v8.3.1 (b9eacd93e9)
- Data source type & version: Elastic / Prometheus, can look up if it matters
- OS Grafana is installed on: Can ask if it matters
- User OS & Browser: Windows 10, Chrome 96
- Grafana plugins: none
- Others:

**Comment by torkelo** [comment]
Any custom plugins installed?

**Comment by Yogu** [comment]
No, no custom plugins.

Seems to only occur with Elasticsearch datasource variables btw

**Comment by Yogu** [comment]
Sorry, we actually have the piechart-panel plugin installed.

**Comment by cykl** [comment]
I just faced the same issue in one dashboard after migrating from 8.2.x to 8.3.2. Thanks @Yogu for providing the two ways to fix it, it worked for me too.

Most of our dashboard seems fine, only one specific dashboard didn't load. This dashboard uses a Prometheus data source and defines interval / constant / query variables. I use Firefox 94. 

Visual diff is messy but here is an excerpt

![image](https://user-images.githubusercontent.com/1764725/146261727-bb569524-bb89-49f6-a868-49cef1234bf9.png)

![image](https://user-images.githubusercontent.com/1764725/146261771-ee7f2caf-29dd-4881-a5c0-f5c8610965bb.png)

I don't know if this help. Please let me know if I can provide useful information to help troubleshooting.

**Comment by codestalkerr** [comment]
I am also facing the same issue after upgrading from 8.2.3 to the 8.3.2 version. Stuffs are broken everywhere and I have many data sources added so hard for me to point something out. 
But thank you @Yogu for the detailing here. Also, I tried what you have mentioned but didn't work for me, so can anyone please help with any workaround, or I should degrade for now?

Note: One point from my end would be I am using mostly Victoria Metrics data source in Grafana.

**Comment by cykl** [comment]
> Most of our dashboard seems fine, only one specific dashboard didn't load. This dashboard uses a Prometheus data source and defines interval / constant / query variables. I use Firefox 94.

I correct this statement. It impacted many of our dashboards (> 20) but not all. Forcing a save always fixed the error. Unfortunately, I haven't been able to identify a plausible criterion that would trigger this issue.

**Comment by ashharrison90** [comment]
struggling to recreate this with any of the provided dashboards or with a simple test case myself locally.

does anyone have a minimal recreation? 🙏

**Comment by parnigot** [comment]
I'm getting the same error on some dashboards after migrating from 8.1 to 8.3.
From my limited investigation the error is somewhat linked to Datasources variables.

In my case all the impacted dashboards used a "complex" set of variables with the following dependencies:

1. A custom variable `Environment` with human readable values like `prod`, `dev`, `uat`
2. A (hidden) `Datasource` variable that used in the `Instance name filter` the `$Environment` to filter down only the correct datasource (I don't know it it matters but in my case I'm using only Elasticsearch datasources)
3. An adhoc filter that was set on the `${Datasource}` variable from above.

With the update/migration to 8.3:

* I'm getting the `Templating init failed Minified Redux...` error on dashboard load
* Ad hoc filters just disappeared from ALL these dashboards
* The `Environment` variable is broken - if I change its value it doesn't update the Datasource correctly

Other dashboards without these variables/dependencies were migrated correctly and work fine in the new version.

After playing a little bit I've found a partial workaround: I can remove the error by manually deleting and recreating from scratch the two remaining variables (`Environment`+`Datasource`).

I say partial workaround because, and maybe this is the root cause, in 8.3 I can't re-create an `Adhoc filter` pointing to a Datasource variable.
I've even tried to do this on a new dashboard and it's just not possible: In 8.1 it was possible to select `${Datasource}` from the drop-down menu but now it doesn't even show up (I'll try to open a dedicated issue on this tonight)

**Comment by cykl** [comment]
>  From my limited investigation the error is somewhat linked to Datasources variables.

Many of our dashboards use data source variable. I'm cannot confirm causality, but this my be a good lead.

**Comment by mxalis** [comment]
The same issue happened for us. It only affected dashboards that used a datasource variable. Other dashboards were not affected.

**Comment by ashharrison90** [comment]
ok finally was able to repro this error with a datasource variable **and** an ad-hoc filter. appreciate all the information from everyone ❤️ 

i believe this is semi related to https://github.com/grafana/grafana/issues/43472 and should be fixed in 8.3.4 by https://github.com/grafana/grafana/pull/43263. if anyone's still seeing this after 8.3.4, please leave a comment and we can reopen.
